### PR TITLE
feat(testing): add public testing surface under nemoguardrails.testing

### PR DESCRIPTION
**Pull request by Pouyanpi**
## Description

Promote FakeLLMModel and TestChat from tests/utils.py to a public nemoguardrails.testing subpackage so downstream users can test their guardrails configurations without copying internal test helpers (dogfooding).

- nemoguardrails.testing.fake_model.FakeLLMModel: framework-agnostic fake LLM that implements the LLMModel protocol.
- nemoguardrails.testing.test_chat.TestChat: ergonomic helper for user/bot conversation assertions.
- nemoguardrails.testing.fixtures: opt-in pytest plugin exposing fake_llm, make_fake_llm, and make_test_chat fixtures.

tests/utils.py is reduced to a thin compatibility shim that re-exports from the new package, so the 100+ existing tests that import from there continue to work unchanged.

Adds tests/testing/test_public_surface.py covering the new exports and fixtures, plus docs/user-guides/testing-your-config.md describing the three supported usage patterns.

## Related Issue(s)

testing subpackage referenced in https://github.com/NVIDIA-NeMo/Guardrails/pull/1857

## Release Notes

* **New Features**
  * Introduced testing utilities for NeMo Guardrails configurations, including deterministic fake LLM support and ergonomic multi-turn conversation testing.
  * Added pytest fixtures and plugin integration for seamless test setup and assertions.

* **Documentation**
  * Added comprehensive user guide on testing configurations, covering error handling, streaming validation, structured outputs, and fixture-based testing approaches.

### CODE REVIEW

File: nemoguardrails/testing/chat_harness.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                                "UtteranceBotActionStarted",
+                                action_uid=event["action_uid"],
+                            )
+                        )
+                        self.input_events.append(
+                            new_event_dict(
+                                "UtteranceBotActionStarted",
+                                action_uid=event["action_uid"],
+                            )
+                        )
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/testing/chat_harness.py:168]: defer + reason: Colang 2.x verification needed before changing

File: nemoguardrails/testing/fake_model.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        response = copy.copy(self._next_response())
+        usage = self._get_usage()
+        if usage:
+            response.usage = usage
+        return response
+
+    async def stream_async(self, prompt, *, stop=None, **kwargs):
+        response = self._next_response()
+        text = response.content
+        chunks = text.split(" ")
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/testing/fake_model.py:117]: skip + reason: type is strict str

File: nemoguardrails/testing/test_chat.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+_TEST_PROVIDERS_WITH_TOKEN_USAGE = ["openai", "azure_openai", "nim"]
+
+
+def _is_data_in_events(events: List[Dict[str, Any]], event_data: List[Dict[str, Any]]) -> bool:
+    if len(events) != len(event_data):
+        return False
+
+    for event, data in zip(events, event_data):
+        if not (all(key in event for key in data) and all(data[key] == event[key] for key in data)):
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/testing/test_chat.py]: thank, made the change:+1 

but these are all pre-existing 👍🏻

File: nemoguardrails/testing/test_chat.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+    """
+
+    __test__ = False
+
+    def __init__(
+        self,
+        config: Union[str, RailsConfig],
+        llm_completions: Optional[List[str]] = None,
+        streaming: bool = False,
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/testing/test_chat.py]: changed the type, it shouldn't be str

File: nemoguardrails/testing/fake_model.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        for chunk_index, chunk in enumerate(chunks):
+            content = chunk + " " if chunk_index < len(chunks) - 1 else chunk
+            await asyncio.sleep(0.05)
+            yield LLMResponseChunk(delta_content=content)
+        await asyncio.sleep(0)
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/testing/fake_model.py:124]: why we were not using 0 ? tests pass after the change :0

File: nemoguardrails/testing/chat_harness.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+            self.llm = FakeLLMModel(
+                responses=llm_completions,
+                streaming=streaming,
+                exception=llm_exception,
+                token_usage=token_usage,
+                should_return_token_usage=should_return_token_usage,
+            )
+        else:
+            self.llm = None
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/testing/chat_harness.py:95]: addressed in the last commit

File: nemoguardrails/testing/fake_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/testing/fake_model.py:117]: addressed in the last commit

File: nemoguardrails/testing/fixtures.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            chat.user("hi")
+            chat.bot("Hi there!")
+    """
+
+    def _factory(
+        config: Union[str, RailsConfig],
+        llm_completions: Optional[List[str]] = None,
+        **kwargs,
+    ) -> TestChat:
+        return TestChat(config=config, llm_completions=llm_completions, **kwargs)
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/testing/fixtures.py:90]: addressed in the last commit

File: nemoguardrails/testing/fixtures.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+        def test_greeting(make_test_chat):
+            config = RailsConfig.from_path("./examples/bots/abc")
+            chat = make_test_chat(config, llm_completions=["Hi there!"])
+            chat.user("hi")
+            chat.bot("Hi there!")
+    """
+
+    def _factory(
+        config: Union[str, RailsConfig],
```
Comment (tgasser-nv) [inline nemoguardrails/testing/fixtures.py]: The `config` here has type of `Union[str, RailsConfig]`, but it's passed into the TestChat init which expects a `RailsConfig` only. If you want to support string configs then it needs to be detected and converted to a RailsConfig before creating the TestChat

File: nemoguardrails/testing/chat_harness.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+                "Hello! How can I help you today?",
+            ],
+        )
+
+        chat.user("Hello! How are you?")
+        chat.bot("Hello! How can I help you today?")
+
+    """
+
+    __test__ = False
```
Comment (tgasser-nv) [inline nemoguardrails/testing/chat_harness.py:57]: I know this removes the file from pytest automatic test discovery. But anyone using this would expect a `test_chat.py` file to contain unit tests for `chat`. Recommend renaming to anything without a Test suffix, ChatHarness, FakeChat maybe? (for symmetry with FakeLLM?)

File: nemoguardrails/testing/fake_model.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+        should_return_token_usage: bool = False,
+    ):
+        if llm_responses is not None:
+            self._llm_responses = llm_responses
+        elif responses is not None:
+            self._llm_responses = [LLMResponse(content=response) for response in responses]
+        else:
+            self._llm_responses = []
+        self.responses = responses or [response.content for response in self._llm_responses]
+        self.i = 0
```
Comment (tgasser-nv) [inline nemoguardrails/testing/fake_model.py]: Can you rename `i` to something more descriptive like `inference_count` or similar? On a public interface it's not obvious what this is counting from the name alone

File: nemoguardrails/testing/chat_harness.py [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+
+    def __init__(
+        self,
+        config: RailsConfig,
+        llm_completions: Optional[List[str]] = None,
+        streaming: bool = False,
+        llm_exception: Optional[Exception] = None,
```
Comment (tgasser-nv) [inline nemoguardrails/testing/chat_harness.py:64]: nit: Should this be renamed `exception` to match the `fake_model.py` `exception field? They're both the same type and it gets passed into the FakeLLMModel init as-is without any changes

File: nemoguardrails/testing/test_chat.py [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    __test__ = False
+
+    def __init__(
+        self,
+        config: RailsConfig,
+        llm_completions: Optional[List[str]] = None,
+        streaming: bool = False,
+        llm_exception: Optional[Exception] = None,
+        token_usage: Optional[List[Dict[str, int]]] = None,
+        llm: Optional[Any] = None,
```
Comment (tgasser-nv) [inline nemoguardrails/testing/test_chat.py]: Should this be type `FakeLLMModel` rather than `Any`? It gets assigned to `self.llm` which has type FakeLLModel on line 88

File: nemoguardrails/testing/fake_model.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        self.streaming = streaming
```
Comment (tgasser-nv) [inline nemoguardrails/testing/fake_model.py]: We don't check this in `stream_async()` below, or use it in any methods. Recommend  removing `self.streaming` and letting the client call either `generate_async()` or `stream_async()` as it needs. Alternatively check this value in `stream_async()` and raise if it's not True

File: nemoguardrails/testing/fake_model.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+
+    async def stream_async(self, prompt, *, stop=None, **kwargs):
+        response = self._next_response()
+        text = response.content or ""
+        chunks = text.split(" ")
+        for chunk_index, chunk in enumerate(chunks):
+            content = chunk + " " if chunk_index < len(chunks) - 1 else chunk
+            await asyncio.sleep(0)
+            yield LLMResponseChunk(delta_content=content)
+        await asyncio.sleep(0)
```
Comment (tgasser-nv) [inline nemoguardrails/testing/fake_model.py:124]: Could you re-add the comment about why the final sleep is needed (from [here](https://github.com/NVIDIA-NeMo/Guardrails/blob/develop/tests/utils.py#L115-L116))?

File: nemoguardrails/testing/__init__.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+Pytest fixtures are exposed via the ``nemoguardrails.testing.fixtures`` plugin.
+Add it to your ``conftest.py`` to opt in::
+
+    pytest_plugins = ["nemoguardrails.testing.fixtures"]
+"""
+
+from nemoguardrails.testing.fake_model import FakeLLMModel
+from nemoguardrails.testing.test_chat import TestChat
+
+__all__ = ["FakeLLMModel", "TestChat"]
```
Comment (tgasser-nv) [inline nemoguardrails/testing/__init__.py:35]: Could you add the other fixtures here too, so import * works? (["fake_llm", "make_fake_llm", "make_test_chat"])

File: nemoguardrails/testing/chat_harness.py [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+            llm_exception: An exception to be raised by the LLM (for testing
+                error handling).
+            token_usage: Optional token usage data for simulating token usage
+                reporting.
+            llm: An explicit LLM instance to bind to the rails app. When
+                provided, ``llm_completions`` is ignored.
+        """
+        if llm is not None:
+            self.llm = llm
+        elif llm_completions is not None or llm_exception is not None:
```
Comment (tgasser-nv) [inline nemoguardrails/testing/chat_harness.py:84]: Could you add a unit-test where llm_completions is None but llm_exception isn't to check the second part of this `elif`?

File: tests/testing/test_public_surface.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    assert first.content == "first"
+    assert second.content == "second"
+
+    with pytest.raises(RuntimeError, match="No responses available"):
+        await llm.generate_async(prompt="q3")
+
+
+@pytest.mark.asyncio
+async def test_fake_llm_model_streaming_yields_chunks():
+    llm = FakeLLMModel(responses=["one two three"])
```
Comment (tgasser-nv) [inline tests/testing/test_public_surface.py:79]: This `llm` object has streaming set to False (the default). As in earlier comment, either remove the streaming arg and let clients call `stream_async()` or `generate_async()` however they want, or keep it and raise if they try to stream when it's not enabled

File: tests/testing/test_public_surface.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (tgasser-nv) [inline tests/testing/test_public_surface.py:1]: nit: Is there a better name for this file? Public Surface doesn't make much sense. Maybe `test_fake_fixtures.py` (hints at FakeLLM or FakeChat if it's renamed to that) or something similar

**Review by tgasser-nv — APPROVED** [review]
Looks good! Mostly nits and cleanups with some renaming needed before merging. 

At a higher-level, what's the purpose of making these test fixtures more publicly accessible? Is it for external contributors to write their own `LLMModel`-compliant classes?

File: nemoguardrails/testing/__init__.py (hunk 15 above)
Comment (Pouyanpi) [reply nemoguardrails/testing/__init__.py:35]: The fixtures fake_llm, make_fake_llm, make_test_chat live in nemoguardrails/testing/fixtures.py as pytest fixtures. the existing __all__  matches what's actually exposed at module level. Users get the fixtures through pytest_plugins = ["nemoguardrails.testing.fixtures"], not through * import.

File: nemoguardrails/testing/fake_model.py (hunk 13 above)
Comment (Pouyanpi) [reply nemoguardrails/testing/fake_model.py]: Yes, good time to clean it up here 👍🏻

File: nemoguardrails/testing/test_chat.py (hunk 12 above)
Comment (Pouyanpi) [reply nemoguardrails/testing/test_chat.py]: 100

File: nemoguardrails/testing/fake_model.py (hunk 10 above)
Comment (Pouyanpi) [reply nemoguardrails/testing/fake_model.py]: I agree, much better name 👍🏻

File: tests/testing/test_public_surface.py (hunk 17 above)
Comment (Pouyanpi) [reply tests/testing/test_public_surface.py:79]: done 👍🏻

File: tests/testing/test_public_surface.py (hunk 18 above)
Comment (Pouyanpi) [reply tests/testing/test_public_surface.py:1]: Skipping this rename for now. the file tests three concerns: the re-exports from nemoguardrails.testing, the tests.utils shim compatibility, and the fixtures. test_fake_fixtures.py would imply only fixtures; the current name covers the broader surface (the public face of the nemoguardrails.testing subpackage). Open to a better suggestion if you have one.

File: nemoguardrails/testing/chat_harness.py (hunk 16 above)
Comment (Pouyanpi) [reply nemoguardrails/testing/chat_harness.py:84]: yes added in fd96d010b61cd495e1c83b92f9873917d96ba7f1

File: nemoguardrails/testing/chat_harness.py (hunk 11 above)
Comment (Pouyanpi) [reply nemoguardrails/testing/chat_harness.py:64]: Nice 👍🏻

File: nemoguardrails/testing/chat_harness.py (hunk 9 above)
Comment (Pouyanpi) [reply nemoguardrails/testing/chat_harness.py:57]: Agreed, i renamed it 👍🏻

File: docs/user-guides/testing-your-config.md [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+    assert result["content"] == "Hello from the fake!"
+```
+
+If you want a regression alarm on prompt changes that introduce extra LLM
+calls, `FakeLLMModel` exposes an `i` counter (the index of the next response,
+which doubles as the count of consumed responses). Most tests assert on the
+response content; the counter is there if you need it:
+
+```python
+assert fake.i == 1, "Expected exactly one LLM call"
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply docs/user-guides/testing-your-config.md]: Valid finding, addressed it 👍🏻

File: docs/user-guides/testing-your-config.md [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+```
+
+If you want a regression alarm on prompt changes that introduce extra LLM
+calls, `FakeLLMModel` exposes an `i` counter (the index of the next response,
+which doubles as the count of consumed responses). Most tests assert on the
+response content; the counter is there if you need it:
+
+```python
+assert fake.inference_count == 1, "Expected exactly one LLM call"
+```
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply docs/user-guides/testing-your-config.md:108]: addressed it in the last commit 👍🏻
